src/block/Hopper.php

```diff
@@ -92,8 +99,231 @@ public function onInteract(Item $item, int $face, Vector3 $clickVector, ?Player
+					$slotInFurnace = FurnaceInventory::SLOT_FUEL;
+					$itemInFurnace = $destination->getInventory()->getFuel();
+				}
+				if(!$itemInFurnace->isNull()){
+					if($itemInFurnace->getCount() >= $itemInFurnace->getMaxStackSize()){
+						return false;
+					}
+					if(!$itemInFurnace->canStackWith($item)){
+						continue;
+					}
+					$item->pop();
+					$itemInFurnace->setCount($itemInFurnace->getCount() + 1);
+				}else{
+					$itemInFurnace = $item->pop();
+				}
+
+				//TODO: event on item inventory switch
+
+				$destination->getInventory()->setItem($slotInFurnace, $itemInFurnace);
+				$inventory->setItem($slot, $item);
+				return true;
+
+			}elseif($destination instanceof TileHopper){
+				$itemToPush = $item->pop();
+				if(!$destination->getInventory()->canAddItem($itemToPush)){
+					continue;
+				}
+				// Hoppers pushing into empty hoppers set the empty hoppers transfer cooldown back to the default amount of ticks.
+				if(count($destination->getInventory()->getContents()) === 0){
+					$destination->setTransferCooldown(TileHopper::DEFAULT_TRANSFER_COOLDOWN);
+				}
+
+			}elseif($destination instanceof \pocketmine\block\tile\Jukebox){
+				if(!($item instanceof Record)){
+					continue;
+				}
+				//TODO:
+				// Jukeboxes actually emit a redstone signal when playing a record so nearby hoppers are blocked and
+				// prevented from inserting another disk. Because neither does redstone work properly nor can we check if
+				// a jukebox is still playing a record or has already finished it, we can just check if it has already a
+				// record inserted.
+				if($destination->getRecord() !== null){
+					return false;
+				}
+
+				// The Jukebox block is handling the playing of records, so we need to get it here and can't use TileJukebox::setRecord().
+				$jukeboxBlock = $destination->getBlock();
+				if($jukeboxBlock instanceof Jukebox){
+					$jukeboxBlock->insertRecord($item->pop());
+					$jukeboxBlock->getPosition()->getWorld()->setBlock($jukeboxBlock->getPosition(), $jukeboxBlock);
+					$inventory->setItem($slot, $item);
+					return true;
+				}
+				return false;
+
+			}elseif($destination instanceof Container){
+				$itemToPush = $item->pop();
+				if(!$destination->getInventory()->canAddItem($itemToPush)){
+					continue;
+				}
+
+			}else{
+				return false;
+			}
+
+			//TODO: event on item inventory switch
+
+			$inventory->setItem($slot, $item);
+			$destination->getInventory()->addItem($itemToPush);
+			return true;
+		}
+		return false;
 	}
 
-	//TODO: redstone logic, sucking logic
+	/**
+	 * This function handles pulling items by the hopper from a container above.
+	 * Returns true if an item was successfully pulled or false on failure.
+	 */
+	private function pull(HopperInventory $inventory, Inventory $origin) : bool{
+		// Hoppers interact differently when pulling from different kinds of tiles.
+		//TODO: Composter
+		//TODO: Brewing Stand
+		//TODO: Jukebox
+		if($origin instanceof FurnaceInventory){
```

Comment:
Each pull logic should be implemented in the different tiles to allow greater flexibility / addition in the future.

Reply:
Same argument as for the push logic.